### PR TITLE
 CDRIVER-4489 refactor URI auth finalization by authentication mechanism

### PR DESCRIPTION
**Pull request by eramongodb**
## Summary

Precursor changes to CDRIVER-4489.

This PR does not implement any `MONGODB-OIDC` features yet. Instead, this PR updates existing code related to authentication mechanisms to better align with the current Drivers [Authentication](https://github.com/mongodb/specifications/blob/b643abfb05c9b8c799723876bba976d9bc05e555/source/auth/auth.md) specification and better faciliate the addition of `MONGODB-OIDC` features. This includes some API breaking changes to correct behavior to conform to the current Authentication spec.

## `MongoCredential`

The Authentication spec is written in terms of a [`MongoCredential`](https://github.com/mongodb/specifications/blob/b643abfb05c9b8c799723876bba976d9bc05e555/source/auth/auth.md#mongocredential) type. The mongoc library has no such type. Instead, the fields are split across data members of `mongoc_uri_t`. Comments have been added to better document the correlation between the `MongoCredential` spec and how they are represented by data members of `mongoc_uri_t`.

Although refactoring the data members to better reflect the `MongoCredential` type specification was considered (e.g. adding a new `mongoc_credential_t` type, moving `MongoCredential.source` into its own data member, etc.), this was rejected due to requiring significant breaking changes to `mongoc_uri_get_credentials` and related API.

## Default `authSource` Behavior

The `mongoc_uri_get_auth_source` function is refactored to better express how defaults are applied for each authentication mechanism. Notably there are four distinct cases to consider:

- "Default Authentication Method": the authentication mechanism to use has not yet been determined (by `_mongoc_cluster_auth_node` during handshake), therefore `authMechanism` is still unset. In this case, the "database or `"admin"`" behavior is applied. This case is deliberately treated separately from explicit authentication mechanism cases for future maintainability (i.e. if `MONGODB-OIDC` is added to the default authentication method).
- "Database or `"admin"`": if `authSource` is not specified, either the database name is used as the source (if provided), or it is defaulted to the `"admin"` database. This applies to `MONGODB-CR`, `MONGODB-SHA-1`, and `MONGODB-SHA-256`.
- "Database or `"$external"`": if `authSource` is not specified, either the database name is used as the source (when provided), or it is defaulted to `"$external"` to indicate an external database must be used for authentication. This only applies to `PLAIN`.
- "Only `$external`": always "`$external`", whether specified or defaulted. This applies to `MONGODB-X509`, `GSSAPI`, and `MONGODB-AWS`.

To avoid significant API breaking changes, validation is _not_ performed by either `mongoc_uri_get_auth_source` or `mongoc_uri_set_auth_source`. However, the return value of `mongoc_uri_get_auth_source` has been fixed to return `"$external"` for `MONGODB-AWS`.

> [!NOTE]
> `mongoc_uri_get_auth_source` may return a source even when `authSource` is not actually present in the URI string itself. This behavior is preserved for backward compatibility.

> [!WARNING]
> **API Breaking Change:** when `authMechanism` is set to `MONGODB-AWS`, `mongoc_uri_get_auth_source` will now default to `"$external"` (correct) instead of "database name or `"admin"` (incorrect).

## URI Authentication Credentials Validation

> [!IMPORTANT]
> All behavior described below applies _only_ to URI finalization during construction of a new `mongoc_uri_t` object from a connection string. It does _not_ apply to other URI-related API (i.e. getters and setters for individual fields and properties).

The internal `mongoc_uri_finalize_auth` function is refactored to better denote how `MongoCredential` properties are validated for each authentication mechanism. The series of _non-exclusive_ `if` statements grouped by validation and defaulting operations are replaced by a series of _exclusive_ `if` statements grouped by authentication mechanism. This refactor allows for a more direct comparison between the spec and the implementation for correctness.

### `authMechanism` Validation

As with `mongoc_uri_get_auth_source`, the "default authentication method" is handled separately from explicitly specified authentication methods. Validation is deferred to `_mongoc_cluster_auth_node` during handshake when the authentication mechanism is determined according to `saslSupportedMechs`. Remaining authentication methods are split into distinct branches by their required `MongoCredential` property requirements.

Even though the default authentication method can (at the moment) only resolve to `SCRAM-SHA-1` and `SCRAM-SHA-256`, the `MongoCredential` property requirements must not be validated "in-advance" of the actual handshake and `saslSupportedMechs` (e.g. the `password` "MUST be specified" requirement). This is both for [backward compatibility](https://github.com/mongodb/specifications/blob/d4e9daae8b1ab7836a9fa5c24f9f2f5db83deaef/source/connection-string/tests/valid-auth.yml#L131) and for forward compatibility (i.e. if `MONGODB-OIDC`, for which `password` "MUST NOT be specified", is added to the default authentication method).

However, `authMechanism` itself _may_ be validated. Invalid values (including "empty" strings and incorrect casing) are now treated as a client error rather than ignored. This change exposed some typos of `"SCRAM-SHA-1"` as `"SCRAM-SHA1"` in test code which are now fixed.

> [!WARNING]
> **API Breaking Changes:**
> - It is now a client error for `authMechanism` to be specified as an invalid or unsupported value (including case-sensitivity, e.g. `"scram-sha-1"` is not equivalent to `"SCRAM-SHA-1"`).
> - It is now a client error for:
>     - a password to be specified for `MONGODB-X509`.
>     - a password to _not_ be specified for `MONGODB-CR`, `MONGODB-SHA-1`, `MONGODB-SHA-256`, and `PLAIN`.
>     - `authMechanismProperties` to be specified for `MONGODB-CR`, `MONGODB-X509`, `MONGODB-SHA-1`, `MONGODB-SHA-256`, and `PLAIN`.
> - It is now a client error when `authSource` is set to a value that is not `"$external"` for `MONGODB-X509`, `GSSAPI`, and `MONGODB-AWS`.

> [!NOTE]
> Although CDRIVER-1959 demotes the username requirement for `MONGODB-X509` from "MUST" to "SHOULD NOT", backward compatibility and [Connection String spec tests](https://github.com/mongodb/specifications/blob/d4e9daae8b1ab7836a9fa5c24f9f2f5db83deaef/source/connection-string/tests/valid-auth.yml#L192) prevent us from validating this requirement as either a warning or an error. This restriction does not apply to password requirements, as described above.

### `authMechanismProperties` Validation

As part of this refactor, validation of authentication mechanisms and related fields is improved. Invalid and unsupported fields in `authMechanismProperties` (when `authMechanism` is specified) ~~now emit warning messages. For backward compatibility, invalid and unsupported fields are still preserved in the resulting `authMechanismProperties` document.~~ are now reported as client errors.

The `CANONICALIZE_HOST_NAME` property for `GSSAPI` is [not fully supported as it should be](https://github.com/mongodb/specifications/blob/b643abfb05c9b8c799723876bba976d9bc05e555/source/auth/auth.md#gssapi) due to CDRIVER-4128. Although the specification requires support for values such as "none", "forward", and "forwardAndReverse", mongoc currently only treats the legacy boolean value "true" (equivalent to "forwardAndReverse") as a significant value; [all other values are ignored](https://github.com/mongodb/mongo-c-driver/blob/ba380a3ce0ef32053b777d2a90a2f22967f67f95/src/libmongoc/src/mongoc/mongoc-sasl.c#L113) and treated as "false" (equivalent to "none"). Such instances in code are now labeled with a comment referencing CDRIVER-4128. ~For backward compatibility, a warning message is emitted rather than returning an error.~ Unsupported values are now diagnosed as a client error [as required by spec](https://github.com/mongodb/specifications/blob/b643abfb05c9b8c799723876bba976d9bc05e555/source/auth/auth.md#hostname-canonicalization).

> [!WARNING]
> **API Breaking Changes:**
> - When `authMechanism` is specified, invalid or unsupported properties in `authMechanismProperties` for the specified authentication mechanism are now a client error (e.g. `CANONICALIZE_HOST_NAME`).

> [!IMPORTANT]
> **New Behavior:**
> - When `authMechanism` is set to `GSSAPI`, a warning message is now emitted for invalid or unsupported values of `CANONICALIZE_HOST_NAME` during a handshake by `_mongoc_sasl_set_properties` (currently, only `"false"` and `"true"` are supported). This may occur when the authentication mechanism or properties are set _after_ URI finalization, such as via `mongoc_uri_set_mechanism_properties`.

## Test Coverage

Comprehensive test coverage of each and every authentication mechanism is implemented as the new `test_mongoc_uri_auth_mechanisms` test function, replacing`test_mongoc_uri_authmechanismproperties`.

This PR additionally applies modernization of URI and connection string test code to improve the quality of assertion messages when comparing integers, strings, and BSON documents.

Before (string comparison; similar for integers):

```c
BSON_ASSERT (0 == strcmp (a, b));
```
```
example.c:123 test_example(): assertion failed: strcmp (a, b) == 0
```

After (string comparison):

```c
ASSERT_STRCMP (a, b);
```
```
FAIL

Assert Failure:
  "contents of a"
  !=
  "contents of b"
example.c:123  test_example()
```

Before (BSON documents, exact match):

```c
BSON_ASSERT (0 == bson_compare(bson, tmp_bson ("{'a': 1}")));
```
```
example.c:123 test_example(): assertion failed: 0 == bson_compare(bson, tmp_bson ("{'b': 2}"))
```

After (BSON documents, exact match):

```
ASSERT_EQUAL_BSON (tmp_bson ("{'a': 1}", bson);
```
```
test error in: example.c 123:test_example()
BSON unequal:
Expected: { "a" : { "$numberInt" : "1" } }
Got     : { "b" : { "$numberInt" : "2" } }
```

Before (BSON documents, element matches):

```c
BSON_ASSERT (bson_iter_init_find (&iter, bson, "x"));
BSON_ASSERT (BSON_ITER_HOLDS_INT32 (&iter));
BSON_ASSERT (bson_iter_int32 (&iter) == 1);
// More `bson_iter_*` operations...
```
```
example.c:123 test_example(): assertion failed: bson_iter_init_find (&iter, bson, "x")
```

After (BSON documents, element matches):

```c
ASSERT_MATCHES (bson, "{'x': 1}");
```
```
ASSERT_MATCH failed:
document: { "y" : { "$numberInt" : "2" }, "z" : { "$numberInt" : "3" } }
pattern : {"x": 1}
error   : x: not found
example.c:123  test_example()
```

### CODE REVIEW

File: NEWS [hunk 1]
```diff
@@ -8,6 +8,30 @@ Removals:
     names `mongoc.h` or `bson.h` directly. Instead, the names must be prefixed
     with the parent directory: `mongoc/mongoc.h` and `bson/bson.h`.
 
+Changes:
+
+  * Authentication mechanisms and connection strings:
+    * Case sensitive is now respected for _values_ of authentication-related URI options and authentication mechanism properties.
+      * _Keys_ remain case-insensitive, including in mechanism properties. e.g. `authMechanismProperties=service_name:value` is equivalent to `authMechanismProperties=SERVICE_NAME:value`.
+      * e.g. `"authMechanism=scram-sha-1"` does not equal `"authMechanism=SCRAM-SHA-1"`.
```
Comment (kevinAlbs) [inline NEWS]: Suggest preserving the existing case-insensitive comparison to reduce the risk of breaking applications upgrading to 2.0.

There may be precedent with other products. Using mongosh (WireShark indicates SCRAM-SHA-1 is used):
```
% mongosh "mongodb://user:password@localhost/?authMechanism=scram-sha-1" --eval 'db.runCommand({"ping": 1})' --quiet
```

**Review by kevinAlbs — COMMENTED** [review]
I like the direction. Posting initial comments.

**Review by eramongodb — COMMENTED** [review]
Some (but not all) `authMechanismProperties` fields for `MONGODB-AWS` and `GSSAPI` are now validated for client errors after the unskipping of several tests related to CDRIVER-3517, CDRIVER-5811, and CDRIVER-5812, all of which are now addressed by this PR.

Summary of notable changes include:

- Removal of remaining `MONGODB-CR`-specific code. (CDRIVER-5773)
- Restoration of case-insensitivity for `authMechanism` values.
- Validate additional client error situations for `MONGODB-AWS` (CDRIVER-5811) and `GSSAPI` (`CANONICALIZE_HOST_NAME`: CDRIVER-4128, not yet resolved).

**Review by eramongodb — COMMENTED** [review]
> [BSON append failures] might have been a bit much to propagate it up to the user when it is internal representation. Converted this error (+ other potential append failures concerning &uri->credentials) into an assertion.

After further consideration, reverted to reporting these internal BSON failures as client errors rather than assertions. This is due to the potential complexity of the contents of the `mongoc_uri_t::credentials` BSON document. Errors pertaining to updates to this field (including via `mongoc_uri_set_mechanism_properties`) use the common "URI credentials BSON error" phrase to indicate their relationship to the URI credentials BSON document, which may be subsequently inspected or fixed via `mongoc_uri_(get|set)_credentials`.

> For backward compatibility, invalid and unsupported fields are still preserved in the resulting authMechanismProperties document.

After further consideration, opted to return a client error (rather than emit a warning) for invalid or unsupported `authMechanismProperties` fields for better consistency with the Authentication spec (deliberately giving priority over Connection String spec), as is currently done for GSSAPI's `CANONICALIZE_HOST_NAME` property:

> Drivers SHOULD raise an error as early as possible when detecting invalid values in a credential. For instance, if a `mechanism_property` is specified for MONGODB-CR, the driver should raise an error indicating that the property does not apply.

File: src/libmongoc/src/mongoc/mongoc-uri.c [hunk 2]
```diff
@@ -1286,71 +1286,469 @@ mongoc_uri_finalize_tls (mongoc_uri_t *uri, bson_error_t *error)
...
+         // service_name refer to the same property.
+         if (strcasecmp (key, prop->name) == 0) {
+            const bson_type_t type = bson_iter_type (&iter);
+
+            if (type == prop->type) {
+               goto found_match; // Matches both key and type.
+            } else {
+               // Connection String spec: Any invalid Values for a given key MUST be ignored and MUST log a WARN level
+               // message.
+               MONGOC_URI_ERROR (error,
```
Comment (kevinAlbs) [inline src/libmongoc/src/mongoc/mongoc-uri.c:1430]: Returning an error (here and below) contradicts the comments:

> Connection String spec: Any invalid Values for a given key MUST be ignored and MUST log a WARN level message.

Suggest referencing the auth spec as rationale:

> Drivers SHOULD raise an error as early as possible when detecting invalid values in a credential.

Related: DRIVERS-2056 proposes revising the connection string spec requirement to warn.

File: src/libmongoc/src/mongoc/mongoc-uri.c [hunk 3]
```diff
@@ -1286,71 +1286,469 @@ mongoc_uri_finalize_tls (mongoc_uri_t *uri, bson_error_t *error)
...
+
+   return true;
+}
+
+static bool
+_finalize_auth_aws_mechanism_properties (const bson_t *mechanism_properties, bson_error_t *error)
+{
+   BSON_OPTIONAL_PARAM (mechanism_properties);
+   BSON_ASSERT_PARAM (error);
+   BSON_OPTIONAL_PARAM (error);
```
Comment (kevinAlbs) [inline src/libmongoc/src/mongoc/mongoc-uri.c:1477]:
```suggestion
```

File: src/libmongoc/src/mongoc/mongoc-uri.c [hunk 4]
```diff
@@ -1286,71 +1286,469 @@ mongoc_uri_finalize_tls (mongoc_uri_t *uri, bson_error_t *error)
...
+                              "'MONGODB-AWS' authentication mechanism requires AWS_SESSION_TOKEN to be accompanied by "
+                              "a username and a password");
+            goto fail;
+         }
+      }
+   }
+
+   // Invalid or unsupported authentication mechanism.
+   else {
+      MONGOC_URI_ERROR (error, "Unsupported value for \"authMechanism\": \"%s\"", mechanism);
```
Comment (kevinAlbs) [inline src/libmongoc/src/mongoc/mongoc-uri.c:1742]: Suggest listing the supported values to help a user identify a possible typo:

```suggestion
      MONGOC_URI_ERROR (error,
                        "Unsupported value for \"authMechanism\": \"%s\". Supported values: \"SCRAM-SHA-1\", "
                        "\"SCRAM-SHA-256\", \"PLAIN\", \"MONGODB-X509\", \"GSSAPI\", \"MONGODB-AWS\"",
                        mechanism);
```

**Review by kevinAlbs — COMMENTED** [review]
The test improvements are much appreciated.

File: src/libmongoc/tests/test-mongoc-connection-uri.c [hunk 5]
```diff
@@ -218,20 +225,98 @@ run_uri_test (const char *uri_string, bool valid, const bson_t *hosts, const bso
...
+         }
+
+         if ((bson_iter_init_find (&iter, credentials, "mechanism_properties") ||
+              bson_iter_init_find (&iter, credentials, "authmechanismproperties")) &&
+             BSON_ITER_HOLDS_DOCUMENT (&iter)) {
+            ASSERT (bson_append_iter (&expected, "authmechanismproperties", -1, &iter));
+         }
+      }
+
+      if (bson_iter_init_find (&iter, &expected, "username") && BSON_ITER_HOLDS_UTF8 (&iter)) {
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/libmongoc/tests/test-mongoc-connection-uri.c]: Updated to use `bsonVisitEach` + `case(when())`. Unexpected fields now trigger a test error. `bsonAs(cstr)` returns `NULL` for incorrect types (i.e. `null`) which now ensures we also test negative assertions (properties/fields which should _not_ be present).

File: src/libmongoc/src/mongoc/mongoc-uri.c [hunk 6]
```diff
@@ -1286,71 +1286,469 @@ mongoc_uri_finalize_tls (mongoc_uri_t *uri, bson_error_t *error)
...
+      }
+
+      // mechanism_properties are allowed for MONGODB-AWS.
+      if (!_finalize_auth_aws_mechanism_properties (mechanism_properties, error)) {
+         goto fail;
+      }
+
+      // Authentication spec: if a username is provided without a password (or vice-versa), Drivers MUST raise an error.
+      if (!username != !password) {
+         MONGOC_URI_ERROR (error, "'%s' authentication mechanism requires both a username and a password", mechanism);
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/libmongoc/src/mongoc/mongoc-uri.c]: Updated wording to: "does not accept a username or a password without the other".

File: src/libmongoc/tests/test-mongoc-uri.c [hunk 7]
```diff
@@ -283,154 +259,1020 @@ test_mongoc_uri_new (void)
...
+
+      // Empty.
+      {
+         mongoc_uri_t *const uri =
+            mongoc_uri_new_with_error ("mongodb://user:@localhost/?" MONGOC_URI_AUTHMECHANISM "=MONGODB-AWS", &error);
+         ASSERT_NO_CAPTURED_LOGS ("mongoc_uri_new_with_error");
+         ASSERT_ERROR_CONTAINS (error,
+                                MONGOC_ERROR_COMMAND,
+                                MONGOC_ERROR_COMMAND_INVALID_ARG,
+                                "'MONGODB-AWS' authentication mechanism requires both a username and a password");
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/libmongoc/tests/test-mongoc-uri.c]: Good catch. Reduced scope of `bson_error_t` objects to minimal scopes to avoid inadvertent reuse of error codes across assertions.

File: src/libmongoc/src/mongoc/mongoc-uri.c [hunk 8]
```diff
@@ -1286,71 +1286,479 @@ mongoc_uri_finalize_tls (mongoc_uri_t *uri, bson_error_t *error)
...
-      /* MONGODB-X509 and MONGODB-AWS are the only mechanisms that don't require
-       * username */
-      if (!(strcasecmp (mongoc_uri_get_auth_mechanism (uri), "MONGODB-X509") == 0 ||
-            strcasecmp (mongoc_uri_get_auth_mechanism (uri), "MONGODB-AWS") == 0)) {
-         if (!mongoc_uri_get_username (uri) || strcmp (mongoc_uri_get_username (uri), "") == 0) {
-            MONGOC_URI_ERROR (
-               error, "'%s' authentication mechanism requires username", mongoc_uri_get_auth_mechanism (uri));
-            return false;
+
+      // Connection String spec: any invalid Values for a given key MUST be ignored and MUST log a WARN level message.
```
Comment (kevinAlbs) [inline src/libmongoc/src/mongoc/mongoc-uri.c:1445]:
```suggestion
      // Authentication spec: Drivers SHOULD raise an error as early as possible when detecting invalid values in a
      // credential. For instance, if a mechanism_property is specified for MONGODB-CR, the driver should raise an error
      // indicating that the property does not apply.
      //
      // Note: this overrides the Connection String spec: Any invalid Values for a given key MUST be ignored and MUST
      // log a WARN level message.

**Review by kevinAlbs — APPROVED** [review]

**Comment by eramongodb** [comment]
> `bsonAs(cstr)` returns `NULL` for incorrect types (i.e. `null`) which now ensures we also test negative assertions (properties/fields which should not be present).

It turns out this is not spec compliant per CDRIVER-5776. Null fields (specified by `~` in YAML or `null` in JSON) are expected to _skip_ relevant assertions, not assert for the absence of said fields. The connection string spec test runner has been updated accordingly, i.e. from the Connection String spec:

> If a test case includes a null value for one of these keys (e.g. auth: `~`, port: `~`), no assertion is necessary. This both simplifies parsing of the test files (keys should always exist) and allows flexibility for drivers that might substitute default values during parsing (e.g. omitted port could be parsed as 27017).

The "keys should always exist" condition is not upheld by the connection string spec test runner due to supporting both connection string, URI, _and_ auth spec tests, which have different field requirements (hosts+auth+options: connection string and URI; credentials: auth).

File: src/libmongoc/src/mongoc/mongoc-uri.c [hunk 9]
```diff
@@ -1286,71 +1286,484 @@ mongoc_uri_finalize_tls (mongoc_uri_t *uri, bson_error_t *error)
...
+                              "authentication mechanism property 'SERVICE_NAME' to 'mongodb': %s",
+                              bsonBuildError);
+            goto fail;
+         }
+         if (!mongoc_uri_set_mechanism_properties (uri, &props)) {
+            MONGOC_URI_ERROR (error,
+                              "unexpected URI credentials BSON error when attempting to default 'GSSAPI' "
+                              "authentication mechanism property 'SERVICE_NAME' to 'mongodb': %s",
+                              bsonBuildError);
+            goto fail;
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/libmongoc/src/mongoc/mongoc-uri.c:1688]: Thank you again. Additionally fixed a potential dreference of null `bsonBuildError` given `mongoc_uri_set_mechanism_properties` failure.

File: src/libmongoc/tests/test-mongoc-connection-uri.c [hunk 10]
```diff
@@ -1,6 +1,13 @@
-#include <mongoc/mongoc.h>
+#include <mongoc/mongoc-uri.h>
+
+//
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/libmongoc/tests/test-mongoc-connection-uri.c:3]: This prevents the `<mongoc/mongoc-uri.h>` component header above from being sorted and grouped with other `#include` directives below. This formatting isn't being applied yet, but it will be the case following CDRIVER-5670.

File: src/libmongoc/tests/test-mongoc-connection-uri.c [hunk 11]
```diff
@@ -219,30 +228,142 @@ run_uri_test (const char *uri_string, bool valid, const bson_t *hosts, const bso
...
+         if ((bson_iter_init_find (&iter, credentials, "mechanism") ||
+              bson_iter_init_find (&iter, credentials, "authmechanism")) &&
+             BSON_ITER_HOLDS_UTF8 (&iter)) {
+            BSON_APPEND_UTF8 (&expected, "authmechanism", bson_iter_utf8 (&iter, NULL));
+         }
+
+         if ((bson_iter_init_find (&iter, credentials, "mechanism_properties") ||
+              bson_iter_init_find (&iter, credentials, "authmechanismproperties")) &&
+             BSON_ITER_HOLDS_DOCUMENT (&iter)) {
+            ASSERT (bson_append_iter (&expected, "authmechanismproperties", -1, &iter));
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/libmongoc/tests/test-mongoc-connection-uri.c]: Good catch. Added `bson_has_field` assertions and added the new keys to the "excluding_noinit" arguments as suggested.
